plutus-core/plutus-core/src/PlutusCore/Constant/Typed.hs

```diff
@@ -356,21 +357,26 @@ newtype Opaque term (rep :: GHC.Type) = Opaque
     { unOpaque :: term
     } deriving newtype (Pretty)
 
+-- | Throw an 'UnliftingError' saying that the received argument is not a constant.
+throwNotAConstant
+    :: (MonadError (ErrorWithCause err term) m, AsUnliftingError err)
+    => term -> m r
+throwNotAConstant = throwingWithCause _UnliftingError "Not a constant" . Just
+
 class AsConstant term where
-    -- | Unwrap a shallowly embedded Haskell value from a @term@ or fail.
-    asConstant :: term -> Maybe (Some (ValueOf (UniOf term)))
+    -- | Unlift from the 'Constant' constructor throwing an 'UnliftingError' if the provided @term@
+    -- is not a 'Constant'.
+    asConstant
+        :: (MonadError (ErrorWithCause err term) m, AsUnliftingError err)
+        => term -> m (Some (ValueOf (UniOf term)))
```

Comment:
This cuts that `Maybe` from the pipeline.

Reply:
Yeah, although we're passing more dictionaries now. They might get inlined, though.